[user triton] add support for prune_configs_by in @triton.autotune

SamGinzburg opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142207


This PR adds support for prune_configs_by in the @triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

SamGinzburg committed:
[inductor] add support for prune_configs_by

[ghstack-poisoned]

SamGinzburg committed:
Update on "[inductor] add support for prune_configs_by"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[inductor] add support for prune_configs_by"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[inductor] add support for prune_configs_by"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1175,36 @@ def call_triton_kernel(
             )
 
+        # Run prune_configs_by to filter the configs
+        if isinstance(variable.kernel, Autotuner):
```
Can we check that either of the config pruning callbacks is actually set before we do the below?

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,71 @@ def f(x, increment):
+            records["run_perf_model"] = True
+            return kwargs["BLOCK_SIZE"]
+
+        configs = [
+            triton.Config(kwargs={"BLOCK_SIZE": 32}),
+            triton.Config(kwargs={"BLOCK_SIZE": 128}),
+        ]
```
Nit: inline this to keep things close?

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,71 @@ def f(x, increment):
+        else:
+            prune_configs_by = {"early_config_prune": early_config_prune}
+
+        @triton.autotune(configs=configs, key=["N"], prune_configs_by=prune_configs_by)
```
`key` is ignored by the current implementation in PT2, so better just leave out to avoid confusion.

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,71 @@ def f(x, increment):
+            x = tl.load(src + offsets, mask=offsets < N)
+            tl.store(dst + offsets, x, mask=offsets < N)
+
+        grid = lambda META: (triton.cdiv(N, META["BLOCK_SIZE"]),)
```
Nit: let's put this inside `f`, as that's normal use case.

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,71 @@ def f(x, increment):
+    def test_triton_kernel_prune_configs_by(
+        self, backend, autotune_at_compile_time, with_perf_model
+    ):
+        N = 1024
+        src = torch.randn(N, device=GPU_TYPE)
+        dst = torch.empty(N, device=GPU_TYPE)
```
Nit: can we move these closer to where they are used, for better readability?

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,71 @@ def f(x, increment):
+
+        @torch.compile(fullgraph=True, backend=backend)
+        def f(dst, src, N):
+            prune_by_kernel[grid](dst, src, N=N)
```
Can we also pass a string and float as additional kwargs, to test the unwrapping? Perhaps some other typically used non-Tensor args, too.

aakhundov reviewed (commented):
Thanks @SamGinzburg! Added a few comments.

SamGinzburg commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1175,36 @@ def call_triton_kernel(
+            realizedArgs = zip(
+                variable.kernel.arg_names, [realize_var(arg) for arg in args]
+            )
+            variable.kernel.nargs = dict(realizedArgs)
```
as far as I can tell in Triton OSS this gets reset on each call to warmup(...) and run(...), used for a few funcs (prehook--unsupported, prune_configs, and warmup--unsupported)

I don't see any harm in saving the previous value though in case this gets changed

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -24,6 +24,7 @@
 import torch.utils._pytree as pytree
 from torch import SymInt, Tensor
 from torch._C import DispatchKey
+from torch._dynamo.variables.lazy import LazyVariableTracker, VariableTracker
```
triton_kernel_wrap.py is generic for the Dynamo vs non-strict export cases. Please do not import Dynamo-specific things in this file.

SamGinzburg replied:
I changed the code to check hasattr(var, "realize") instead of importing this, is that the right thing to do here?

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1175,36 @@ def call_triton_kernel(
+            variable.kernel.perf_model = None
+            variable.kernel.configs_top_k = 1.0
+            variable.kernel.early_config_prune = None
+
```
Would be helpful (to other folks looking at this code in the future) if the PR body had some more context

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1175,36 @@ def call_triton_kernel(
+            # args and kwargs aren't real values yet
+            # e.g., LazyVariableTracker
+            # If a user wants to access them, we should realize them
+            def realize_var(
+                var: VariableTracker,
+            ) -> Union[Any, VariableTracker, LazyVariableTracker]:
```
What's the high level strategy here? Are we  "realizing" the configs, then running the prune_configs_by function, then storing that information somewhere?

aakhundov replied:
@zou3519 in this PR, we want to add support for the `prune_configs_by` parameter of the `@triton.autotune` decorator (see the docs [here](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune)). (@SamGinzburg could you please update the PR head with more context?)

The implementation relies on the `Autotuner.prune_configs` method ([here](https://github.com/triton-lang/triton/blob/73ba8b639226fbc505115db59189a1362c350075/python/triton/runtime/autotuner.py#L214)) which runs the user-provided callbacks (`early_config_prune` and `perf_model`) on the `args` and `kwargs` passed to the kernel and prunes the configs in-place (i.e., resets the `Autotuner.configs` attribute to a smaller list). So two important things here are:

- We should set `args` as `Autotuner.nargs` and pass `kwargs` *separately*.

- The original `args` and `kwargs` values (not variable trackers) should be passed to the callbacks, so we should realize before passing.

We've added the logic in the triton_kernel_wrap, because the original `args` and `kwargs` are still separated here. To make the code generic wrt. Dynamo vs non-strict export, we could add a helper method `TritonHOPifier.realize` for realizing the values, which would be implemented differently in the `DynamoTritonHOPifier` and `TracingTritonHOPifier` subclasses. Then we can call `self.realize` in the `TritonHOPifier.call_triton_kernel` and dispatch to the implementation-specific version of realization. Does this make sense to you?

But I also have two questions re. specific implementations of `realize`:

1. Is the Dynamo-specific realization code above valid or are we missing smth / is there a better way?

2. What does realization look like for non-strict export? Is it a no-op (i.e., the values are the same as passed to the kernel call) or something needs to be done for realizaiton?

Thanks!

zou3519 replied:
> Is the Dynamo-specific realization code above valid or are we missing smth / is there a better way?

We usually don't unpack values back into their eager types inside of Dynamo. Also, this is only really OK to do if all of the types are constants (that is, Dynamo guards on their values), and if the `early_config_prune`/`perf_model`/other user-defined functions are "safe to execute" (that is, it has no side effects, which is really not guaranteed here).

The alternative strategy that you probably want to do instead is to create a function in Dynamo and then invoke it with the VariableTracker arguments, so that Dynamo executes the function via its bytecode interpreter (instead of cpython's). Something like the following:
```py
def do_prune_configs(early_config_prune, perf_model, configs_top_k, nargs, configs, kwargs):
    autotuner = object()
    autotuner.early_config_prune = early_config_prune
    autotuner.perf_model = perf_model
    autotuner.configs = configs
    autotuner.configs_top_k = configs_top_k
    configs = AutoTuner.prune_configs(autotuner, kwargs)
    return configs

UserDefinedFunctionVariable(do_prune_configs).call_function(tx, variable_trackers)
```

aakhundov replied:
I'm a bit worried about duck typing of the `Autotuner` in the above: if the implementation details in the Triton code change, this may break. What would be the `variable_trackers` passed to the `UserDefinedFunctionVariable(do_prune_configs).call_function`? Because we have `early_config_prune` and `perf_model` as Python functions on the `Autotuner` object, not variable trackers. If passing normal Python objects to `UserDefinedFunctionVariable(do_prune_configs).call_function` is allowed, can we just pass our `autotuner` object instead and call the `autotuner.prune_cofigs(kwargs)` directly? The call will modify the state of the `autotuner` object, but it is not tracked by Dynamo.

aakhundov replied:
Alternatively, we can try implementing `prune_configs_by` support in Inductor, as we'll have access to the `Autotuner` object there. We'll need to pass the `len(args)` which we can do through patching the `Autotuner` object in Dynamo and de-patching in Inductor.

zou3519 reviewed (changes requested):
Dynamo-specific things shoud not be in triton_kernel_wrap.py

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1175,36 @@ def call_triton_kernel(
+                elif hasattr(var, "get_real_value"):
+                    return var.get_real_value()
+                else:
+                    return var
```
I changed this to raise an error

SamGinzburg commented:
Hey, just wanted to add a note---the dynamo version of this PR seems a bit complicated, we are going to explore implementing this in inductor

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,155 @@ def f(x, increment):
+            records["capture_named_args"] = True
+        return [configs[0]]
+
+    In OSS Triton this user-defined function executes correctly in the test.
+    We execute early_config_prune in dynamo, so `records` is not updated.
```
As discussed offline with @zou3519, Dynamo should support updating the nonlocals / globals, so we should try to fix this.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1195,45 @@ def call_triton_kernel(
                 "Please use a Config in @triton.autotune instead."
             )
 
+        # We support running a single Autotuner for each Triton kernel
+        # Currently, if there are multiple autotuning decorators, the subsequent ones will be silently ignored
+        # We raise an error here to avoid silent incorrectness
+        iter_kernel = variable.kernel
+        autotuner_count = 0
+        while not isinstance(iter_kernel, JITFunction):
+            if isinstance(iter_kernel, Autotuner):
+                autotuner_count += 1
+            if autotuner_count > 1:
+                self.raise_unsupported(
+                    "Passing multiple @triton.autotune decorators is not supported. "
+                    "Please use a single @triton.autotune decorator instead."
+                )
+            iter_kernel = iter_kernel.fn
```
Wondering if we could move erroring on nested autotuners to a separate PR, as it's not strictly required for this one?

SamGinzburg replied:
this is needed for the heuristics PR, but I can put this in a separate PR

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1195,45 @@ def call_triton_kernel(
+                )
+            iter_kernel = iter_kernel.fn
+        # These are the default values in upstream Triton
+        # see: triton/runtime/autotuner.py
```
Better make a permalink with line number, to be specific.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1195,45 @@ def call_triton_kernel(
+        default_configs_top_k = 1.0
+        default_early_config_prune = None
+
+        # Save the arguments length for prune_configs_by
+        # We need this to get the args from kwargs later in get_kernel_and_metadata
```
No longer relevant?

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1195,45 @@ def call_triton_kernel(
+        # These are the default values in upstream Triton
+        # see: triton/runtime/autotuner.py
+        default_perf_model = None
+        default_configs_top_k = 1.0
```
This seems unused.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1195,45 @@ def call_triton_kernel(
+            # Prune the configs
+            named_args = dict(zip(variable.kernel.arg_names, args))
+
+            assert tx is not None
```
Will `tx` not be `None` in non-strict export mode (with `triton_op`)?

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1193,12 +1248,28 @@ def call_triton_kernel(
                     config.__dict__.update(special_kwargs)
-                new_kernel = autotune(configs=new_configs, key=[])(variable.kernel.fn)
+
+                prune_configs_by = {
+                    "perf_model": variable.kernel.perf_model,
+                    "early_config_prune": variable.kernel.early_config_prune,
+                    "configs_top_k": variable.kernel.configs_top_k,
+                }
+                new_kernel = autotune(
+                    configs=new_configs, prune_configs_by=prune_configs_by, key=[]
+                )(variable.kernel.fn)
```
Can we extract this into a helper, as doing multiple times?

SamGinzburg replied:
I think we may not need this anymore, since prune_configs_by should be stripped from the kernel by there (we recurse and run call_triton_kernel with the new configs earlier)

aakhundov commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,83 @@ def call_grid(self, grid, meta, tx):
         return grid
 
+    # We use this function to wrap user-defined functions (e.g., early_config_prune, perf_model)
+    def call_user_defined_fn(self, user_fn, args, kwargs, tx):
+        result = UserFunctionVariable(user_fn).call_function(tx, args, kwargs)
+        return result
```
So it seems that `call_user_defined_fn` is only called from the dynamo-specific implementation of  `call_prune_configs_if_required`. Do we really need to make it an abstract method of `TritonHOPifier`? Or maybe a private helper in the `DynamoTritonHOPifier` suffice?

aakhundov commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,83 @@ def call_grid(self, grid, meta, tx):
+    def call_prune_configs_if_required(  # type: ignore[no-untyped-def]
+        self,
+        autotuner,
+        args,
```
Can we rename `args` to `named_args` here to avoid confusion?

aakhundov commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,83 @@ def call_grid(self, grid, meta, tx):
+                tx,
+            )
+        # realize the pruned configs
+        configs_variable = [config.realize().value for config in configs_variable.items]
```
Let's use `configs` here on the lhs, as it's no longer a dynamo variable?

aakhundov commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,83 @@ def call_grid(self, grid, meta, tx):
+                    # call the user-defined function here as well
+                    config: self.call_user_defined_fn(
+                        autotuner.kernel.perf_model,
+                        [args],
```
I think, `nargs` are also passed as part of kwargs to `perf_model`: https://github.com/triton-lang/triton/blob/d7ebf7982adedb61095b93084dcbcaf1d28c8da4/python/triton/runtime/autotuner.py#L225

aakhundov commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,83 @@ def call_grid(self, grid, meta, tx):
+                    config: self.call_user_defined_fn(
+                        autotuner.kernel.perf_model,
+                        [args],
+                        {**kwargs, **config.all_kwargs()},
```
IIUC, the values in `**config.all_kwargs()` here won't be Dynamo variables. Will this work with the Dynamo-interpreted call of the `autotuner.kernel.perf_model`?

SamGinzburg replied:
We realize each Config here: https://github.com/pytorch/pytorch/pull/142207/files#diff-6cfc97f79ba69f9eb1fd097057e89f00b8a35797b2b8e683d44ac734d491a68cR1104

we could add an assert that this has succeeded

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,82 @@ def call_grid(self, grid, meta, tx):
 
+    # We use this function to wrap user-defined functions (e.g., early_config_prune, perf_model)
+    def call_user_defined_fn(self, user_fn, args, kwargs, tx):
+        result = UserFunctionVariable(user_fn).call_function(tx, args, kwargs)
```
Use VariableBuilder instead of directly calling the UserFunctionVariable ctor

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,82 @@ def call_grid(self, grid, meta, tx):
+                tx,
+            )
+        # realize the pruned configs
+        configs = [config.realize().value for config in configs.items]
```
Use `config.guard_as_python_constant()` instead of realize().value -- the idea is that if the configs object actually changes for some reason, then Dynamo needs to recompile. `guard_as_python_constant` inserts some guards for Dynamo to check if the configs object changed.

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,279 @@ def f(x, increment):
+            self.assertEqual(src + 1.5, dst)
+        else:
+            # We require the largest config to be picked for the correct result (BLOCK_SIZE==128)
+            self.assertNotEqual(src + 1.5, dst)
```
Maybe check what it should be equal instead?

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,279 @@ def f(x, increment):
+        f(dst, src, 1.5, "TEST", True, N)
+
+        if with_perf_model:
+            # Disabled for now, some dynamo weirdness at play here
```
Outdated?

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,279 @@ def f(x, increment):
+            self.assertEqual(len(records), 1)
+            self.assertEqual(src + 1.5, dst)
+        else:
+            # We require the largest config to be picked for the correct result (BLOCK_SIZE==128)
```
Shouldn't early_config_prune pick `[configs[0]]` which is 32? Also the assertion below would fail with 128?

SamGinzburg replied:
We call self.assertNotEqual because 32 is picked in this case (resulting in dst == src here) I can change the test to assertEqual instead and clarify things

aakhundov replied:
Yeah, I mean, the comment seems wrong.

SamGinzburg replied:
return kwargs["BLOCK_SIZE"] * -1 should result in a BLOCK_SIZE of 128 (largest config) since it is the largest one (after *-1 it is treated as the smallest for perf_model)

aakhundov replied:
What you're describing happens in the `perf_model`, but we're in the `else:` branch here? Or am I missing the meaning of the comment?

SamGinzburg replied:
I'll clarify whats going on here in the comments

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,279 @@ def f(x, increment):
+            # We require the largest config to be picked for the correct result (BLOCK_SIZE==128)
+            self.assertNotEqual(src + 1.5, dst)
+            self.assertEqual(len(records), 3)
+            self.assertTrue(records["run_early_config_prune"] is not None)
+            self.assertTrue(records["capture_kwargs"] is not None)
+            self.assertTrue(records["capture_named_args"] is not None)
```
False is also not None. Let's just remove `is not None`?

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,279 @@ def f(x, increment):
+            tl.store(dst + offsets, x, mask=offsets < N)
+
+        torch._dynamo.reset()
+        counter = 0
+
+        # Track how many compiles occur
+        @torch._dynamo.on_compile_start
+        def start_callback():
+            nonlocal counter
+            counter += 1
```
There must be an existing counter for recompilation that we could reuse?

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,279 @@ def f(x, increment):
+
+        self.assertEqual(counter, 1)
+
+        # configs has changed (because of the pruning), so this will recompile
+        f(dst, src, 1.5, "TEST", True, N)
+
+        self.assertEqual(counter, 2)
```
Why do we recompile here? Nothing has changed on the kernel, configs, callbacks, or arguments.

SamGinzburg replied:
this was happening because of the records dict---fixed now

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,279 @@ def f(x, increment):
+    # Test if non-strict export works
+    @requires_gpu
+    @common_utils.parametrize("with_perf_model", [True, False])
+    def test_triton_prune_configs_non_strict_export(self, with_perf_model):
```
Can we merge this test with the first one? Seems that the only difference is the `@torch.library.triton_op` decorator, which we can apply on a boolean flag?

aakhundov reviewed (commented):
Thanks @SamGinzburg, we're getting closer! Left some comments. Will defer to @zou3519 to review the Dynamo-related pieces.

SamGinzburg commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,44 @@ def call_grid(self, grid, meta, tx):
+        )
+        return wrapped_user_obj
+
+    # We need to override call_getitem here so that we can call GetItemSource
```
I ended up just calling variable.source instead. (I think this was for the case where we call the triton kernel with a grid and need to pass the source through?)

SamGinzburg commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1231,12 +1303,100 @@ def call_triton_kernel(
                     new_var = type(variable)(new_kernel, None, variable.grid)
                     return self.call_triton_kernel(new_var, args, kwargs, tx)
 
+        if "num_ctas" in kwargs:
+            self.raise_unsupported(
+                "Passing num_ctas directly to the Triton kernel is not supported. "
+                "Please use a Config in @triton.autotune instead."
+            )
```
the answer to this was kind of interesting.

I had to set variable.source in call_getitem in functions.py in order to pass the source through (we need a valid base source for AttrSource). This is because of how Triton kernels are called with the grid.

It turns out that passing the source through breaks that test, and causes it to fail with a runtime error:  

<img width="1543" alt="image" src="https://github.com/user-attachments/assets/868bcc9f-000f-4432-bf51-50dd23e775bd" />

Moving the check down fixes the test, but I'll think about this some more and see if there is a better way to address this

SamGinzburg commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1231,12 +1303,100 @@ def call_triton_kernel(
+            )
+
+            # The result of prune_configs must be a var sequence
+            # Similar to guard_as_python_constant, this function inserts
+            # guards for Dynamo.
```
upon looking I think this just unpacks the list (return type is ListVariable).

guard_as_python_constant installs the guards, I will update the comment

SamGinzburg commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1231,12 +1303,100 @@ def call_triton_kernel(
+                variable_source,
+            )
+
+            # The result of prune_configs must be a var sequence
+            # Similar to guard_as_python_constant, this function inserts
+            # guards for Dynamo.
+            pruned_configs = pruned_configs.unpack_var_sequence(tx)
+            # guard_as_python_constant inserts some guards for Dynamo to check if the configs object changed.
+            pruned_configs = [
+                config.guard_as_python_constant() for config in pruned_configs
+            ]
```
I added a test for non-strict export mode and it seems to work

(test_triton_prune_configs_non_strict_export)

aakhundov replied:
Can we double check that this test is doing what is intended? From the code (and my understanding of non-strict export), it seems that `config.guard_as_python_constant()` should fail. So I'm wondering if `config` is actually a `triton.Config` in the non-strict export test.

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg commented on torch/_dynamo/trace_rules.py:
```diff
@@ -3294,6 +3296,7 @@ def _module_dir(m: types.ModuleType):
     "torch._higher_order_ops.invoke_subgraph",
     "torch._higher_order_ops.scan",
     "torch._higher_order_ops.strict_mode",
+    "torch._higher_order_ops.triton_kernel_wrap",
```
If we don't allow this function here in trace_rules, then do_prune_configs will get marked as a "SkipFunctionVariable" instead of a UserFunctionVariable---https://github.com/pytorch/pytorch/blob/5e67bea7aa3e256243020e10978bd869d62472ae/torch/_dynamo/trace_rules.py#L71

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3656,6 +3658,185 @@ def add(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+                x = x + add_float
+            tl.store(dst + offsets, x, mask=offsets < N)
+
+        if backend == "non-strict":
+            decorator = torch.library.triton_op(f"{libname}::{opname}", mutates_args={})
```
I think, `non-strict` should be orthogonal to the `backend`. I.e., `non-strict` determines how the Triton kernel is wrapped, whereas the backend is passed to `torch.compile`. When `non-strict` is set, we should wrap the kernel into `triton_op`. And then we should `torch.compile` in any case: whether `non-strict` is enabled or not.

aakhundov replied:
Should `mutated_args` include `dst`, as we modify it?

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1171,17 +1234,22 @@ def call_triton_kernel(
+            self.raise_unsupported("Triton kernels should always be called with a grid")
+
+        """
+        Currently, if there are multiple autotuning decorators, the subsequent ones will be silently ignored.
```
Please remove this comment as confusing. This is already addressed in https://github.com/pytorch/pytorch/pull/143519.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1250,19 +1330,89 @@ def call_triton_kernel(
+                variable,
+            )
+
+            # unpack the list of configs
+            pruned_configs = pruned_configs.unpack_var_sequence(tx)
+
+            # guard_as_python_constant inserts guards for Dynamo to check if the configs object changed.
+            pruned_configs = [
+                config.guard_as_python_constant() for config in pruned_configs
+            ]
```
This still looks like dynamo-specific code. In non-strict export mode, `pruned_configs` will just be `List[triton.Config]`. So methods like `unpack_var_sequence` and `guard_as_python_constant` shouldn't be available.

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3656,6 +3658,187 @@ def add(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+            grid = lambda META: (triton.cdiv(N, META["BLOCK_SIZE"]),)
+            prune_by_kernel[grid](dst, src, add_float, N=N)
+
+        if tracing == "non-strict":
+            decorator = torch.library.triton_op(
+                f"{libname}::{opname}", mutates_args={"dst"}
+            )
+        else:
+            # we can just pass the function 'f' for dynamo
+            decorator = f
```
```suggestion
        if tracing == "non-strict":
            f = torch.library.triton_op(
                f"{libname}::{opname}", mutates_args={"dst"}
            )(f)
```

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3656,6 +3658,189 @@ def add(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+            else:
+                prune_by_kernel[grid](dst, src, add_float, N=N)
+
+        if non_strict:
+            decorator = torch.library.triton_op(
+                f"{libname}::{opname}", mutates_args={"dst"}
+            )(f)
+        else:
+            # we can just pass the function 'f' for dynamo
+            decorator = f
+
+        compiled_f = torch.compile(decorator, backend=backend)
```
Nit: strictly, there is no decorator in this code, so the naming seems a bit confusing.

aakhundov reviewed (approved):
The code LGTM! @zou3519 could you please confirm the Dynamo-related parts?

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1039,6 +1049,52 @@ def call_grid(self, grid, meta, tx):
+    def call_user_defined_fn(self, user_fn, args, kwargs, tx, variable):
+        from .builder import VariableBuilder
+
+        wrapped_user_function = VariableBuilder(tx, variable.source)._wrap(user_fn)
```
Don't use VariableBuilder here -- it looks like the only usage is with `user_fn=do_prune_configs`. The source of this is not variable.source (where variable is the TritonKernelVariable).

I don't think do_prune_configs doesn't need a source. Is it possible to use SourcelessBuilder on this? If not then it is OK to directly create a UserDefinedFunctionVariable

zou3519 reviewed (approved):
Two of the sources look off, let's fix that. Otherwise this looks good to me

SamGinzburg committed:
Update on "[user triton] add support for prune_configs_by in @triton.autotune"



This PR adds support for prune_configs_by in the triton.autotune decorator [docs](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune). Supporting this lets users reduce autotuning time by running user-supplied code (early_config_prune, perf_model) to prune the provided list of configs.

We implement this by realizing args/kwargs in call_triton_kernel(...), and then calling kernel.prune_configs(...).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg commented:
@pytorchbot merge

SamGinzburg commented:
@pytorchbot merge
